pkg/kv/kvserver/batcheval/cmd_gc.go

```diff
@@ -202,10 +204,28 @@ func GC(
 			rk.StartKey, rk.EndKey, cArgs.EvalCtx.GetGCThreshold(), cArgs.EvalCtx.GetMVCCStats()); err != nil {
 			return result.Result{}, err
 		}
+
+		// If all data in range was deleted, we need to reset the hint to prevent
+		// GC from touching this range unnecessarily.
+		sl := MakeStateLoader(cArgs.EvalCtx)
+		hint, err := sl.LoadGCHint(ctx, readWriter)
+		if err != nil {
+			return result.Result{}, err
+		}
+		if !hint.IsEmpty() {
+			hint.ResetLatestRangeDeleteTimestamp()
```

Comment:
This is a good point. So I think we could safely remove hints if gc timestamp set on replica is newer than gc hint timestamp. We can do that regardless of GC operation performed. If we fail to pick up hint or there was a newer data, then we'll just remove it because it is obsolete.
For merge case we were trying to handle case where we merge with neighbouring ranges. As you pointed this is problematic as we could try to extend hint to non eligible neighbours and push it up. I think looking on stats during merge and only extending hint to neightbours is only valid when mvcc stats are not estimates and they report no data in the range, then we can safely merge. Beside the case where hint is set on both halfs of the merge of course, in that case, picking newer timestamp seems valid.

Reply:
Oh I think I missed the point that if we clear it on the set threshold, but not when cleaning then the hing would be actually lost. Let me update that.